src/java.base/share/classes/jdk/internal/foreign/abi/CapturableState.java

```diff
@@ -30,66 +30,79 @@
-    ERRNO             ("errno",           JAVA_INT, 1 << 2, true);
+public final class CapturableState {
 
-    public static final StructLayout LAYOUT = MemoryLayout.structLayout(
-        supportedStates().map(CapturableState::layout).toArray(MemoryLayout[]::new));
-    public static final List<CapturableState> BY_ORDINAL = List.of(values());
+    public static final StructLayout LAYOUT;
+    private static final Map<String, CapturableState> LOOKUP;
 
     static {
-        assert (BY_ORDINAL.size() < Integer.SIZE); // Update LinkerOptions.CaptureCallState
+        List<CapturableState> supported;
+
+        if (OperatingSystem.isWindows()) {
+            supported = List.of(
+                    new CapturableState("GetLastError",    JAVA_INT, 1 << 0),
+                    new CapturableState("WSAGetLastError", JAVA_INT, 1 << 1),
+                    new CapturableState("errno",           JAVA_INT, 1 << 2)
+            );
+        } else {
+            supported = List.of(new CapturableState("errno", JAVA_INT, 1 << 2));
+        }
+
+        MemoryLayout[] stateLayouts = new MemoryLayout[supported.size()];
+        @SuppressWarnings({"unchecked", "rawtypes"})
+        Map.Entry<String, CapturableState>[] entries = new Map.Entry[supported.size()];
+        int i = 0;
+        for (var each : supported) {
+            stateLayouts[i] = each.layout;
+            entries[i] = Map.entry(each.stateName, each);
+            i++;
+        }
+        LAYOUT = MemoryLayout.structLayout(stateLayouts);
+        LOOKUP = Map.ofEntries(entries);
     }
 
-    private final String stateName;
-    private final ValueLayout layout;
-    private final int mask;
-    private final boolean isSupported;
+    public final String stateName;
+    public final ValueLayout layout;
+    public final int mask;
 
-    CapturableState(String stateName, ValueLayout layout, int mask, boolean isSupported) {
+    private CapturableState(String stateName, ValueLayout layout, int mask) {
         this.stateName = stateName;
         this.layout = layout.withName(stateName);
         this.mask = mask;
-        this.isSupported = isSupported;
-    }
-
-    private static Stream<CapturableState> supportedStates() {
-        return Stream.of(values()).filter(CapturableState::isSupported);
     }
 
     public static CapturableState forName(String name) {
-        return Stream.of(values())
-                .filter(stl -> stl.stateName().equals(name))
-                .filter(CapturableState::isSupported)
-                .findAny()
-                .orElseThrow(() -> new IllegalArgumentException(
-                        "Unknown name: " + name +", must be one of: "
-                            + supportedStates()
-                                    .map(CapturableState::stateName)
-                                    .collect(Collectors.joining(", "))));
-    }
-
-    public String stateName() {
-        return stateName;
+        var ret = LOOKUP.get(name);
```

Comment:
As we usually have just one element (and for Windows three elements) it might be both less memory consuming and faster to just search a `@Stable` array linearly rather than having a Map.

Reply:
Alternatively, you could hard code the names in a `switch` here.